pkg/kubelet/status/manager.go

```diff
@@ -211,56 +250,75 @@ func (m *manager) RemoveOrphanedStatuses(podUIDs map[types.UID]bool) {
 }
 
 // syncBatch syncs pods statuses with the apiserver.
-func (m *manager) syncBatch() error {
-	syncRequest := <-m.podStatusChannel
-	pod := syncRequest.pod
-	status := syncRequest.status
-
-	var err error
-	statusPod := &api.Pod{
-		ObjectMeta: pod.ObjectMeta,
+func (m *manager) syncBatch() {
+	m.podStatusesLock.RLock()
+	defer m.podStatusesLock.RUnlock()
+
+	// Clean up orphaned versions.
+	for uid := range m.apiStatusVersions {
+		if _, ok := m.podStatuses[uid]; !ok {
+			delete(m.apiStatusVersions, uid)
+		}
+	}
+
+	for uid, status := range m.podStatuses {
+		if m.needsUpdate(uid, status) {
+			m.syncPod(uid, status)
+		}
+	}
+}
+
+// syncPod syncs the given status with the API server. The caller must not hold the lock.
+func (m *manager) syncPod(uid types.UID, status versionedPodStatus) {
+	if !m.needsUpdate(uid, status) {
+		glog.Warningf("Status is up-to-date; skipping: %q %+v", uid, status)
+		return
 	}
+
 	// TODO: make me easier to express from client code
-	statusPod, err = m.kubeClient.Pods(statusPod.Namespace).Get(statusPod.Name)
+	pod, err := m.kubeClient.Pods(status.podNamespace).Get(status.podName)
 	if errors.IsNotFound(err) {
-		glog.V(3).Infof("Pod %q was deleted on the server", pod.Name)
-		return nil
+		glog.V(3).Infof("Pod %q was deleted on the server", status.podName)
+		return
 	}
 	if err == nil {
-		if len(pod.UID) > 0 && statusPod.UID != pod.UID {
-			glog.V(3).Infof("Pod %q was deleted and then recreated, skipping status update", kubeletutil.FormatPodName(pod))
-			return nil
+		if len(pod.UID) > 0 && pod.UID != uid {
+			glog.V(3).Infof("Pod %q was deleted and then recreated, skipping status update",
+				kubeletutil.FormatPodName(pod))
+			delete(m.apiStatusVersions, uid)
+			return
 		}
-		statusPod.Status = status
+		pod.Status = status.status
 		// TODO: handle conflict as a retry, make that easier too.
-		statusPod, err = m.kubeClient.Pods(pod.Namespace).UpdateStatus(statusPod)
+		pod, err = m.kubeClient.Pods(pod.Namespace).UpdateStatus(pod)
```

Comment:
I'm a little confused by the positioning of this comment - where is "here"? Do you mean above where we handle the NotFound error? And maybe also where we handle the recreated case? We have cleanup loops for both orphaned statuses and apiStatusVersions, but I agree that we should try to minimize the need for cleaning up there.